### PR TITLE
Support arbitrary repositories and branches or commits

### PR DESCRIPTION
**Pull request by mikeharder**
@halter73 @davidfowl 

Example:
```
dotnet run -- -n plaintext -s http://adx-3u-perfsvr:5001 -c http://adx-3u-svr02:5002
  -o KestrelHttpServer:user/test-branch -o HttpAbstractions:user/test-branch
```

### CODE REVIEW

**Comment by BrennanConroy** [comment]
This doesn't look like it supports arbitrary commits, wouldn't that be something we would want?

**Comment by mikeharder** [comment]
@BrennanConroy This is a good suggestion, I will add this in a subsequent PR.  For syntax to the BenchmarkDriver app, I was thinking of using "repo:branch#commit", like so:

```
-o KestrelHttpServer:user/test-branch#a1b2c3d
```

**Comment by rynowak** [comment]
Why not `@` - that's what github does

**Comment by mikeharder** [comment]
OK, then:

```
-o KestrelHttpServer:user/test-branch@a1b2c3d
```

**Comment by mikeharder** [comment]
I chatted with @halter73, we have improved the design.  In the driver, you can specify either a branch or a commit after the '@' character:

```
-o KestrelHttpServer@user/test-branch
-o KestrelHttpServer@a1b2c3d
```

The code can handle these in exactly the same way:

```
git clone repo
git checkout branch|commit
```

I will update this PR with these changes.

File: src/BenchmarksServer/Startup.cs [hunk 1]
```diff
@@ -174,36 +169,46 @@ public static int Main(string[] args)
...
+            {
+                globalJson["projects"].Add(Path.Combine("..", dir, "src"));
+            }
+            File.WriteAllText(globalJsonPath, JsonConvert.SerializeObject(globalJson, Formatting.Indented));
 
-                // Add references to libuv (required until https://github.com/aspnet/KestrelHttpServer/pull/731)
-                var benchmarksProjectJson = Path.Combine(path, _benchmarksDir, "src", "Benchmarks", "project.json");
-                dynamic projectJson = JsonConvert.DeserializeObject(File.ReadAllText(benchmarksProjectJson));
+            // If any dir named "KestrelHttpServer", add libuv reference to Benchmarks
+            // (required until https://github.com/aspnet/KestrelHttpServer/pull/731)
```
Comment (halter73) [inline src/BenchmarksServer/Startup.cs]: No longer required: https://github.com/aspnet/KestrelHttpServer/pull/731

File: src/BenchmarksServer/Startup.cs [hunk 2]
```diff
@@ -174,36 +169,46 @@ public static int Main(string[] args)
             }
         }
 
-        private static void CloneAndRestore(string path, ServerJob job)
+        private static string CloneAndRestore(string path, ServerJob job)
         {
-            var benchmarksRepoUrl = string.IsNullOrEmpty(job.BenchmarksRepo) ? _defaultBenchmarksRepoUrl : job.BenchmarksRepo;
-            var benchmarksBranch = string.IsNullOrEmpty(job.BenchmarksBranch) ? _defaultBranch : job.BenchmarksBranch;
-            Git.Clone(path, benchmarksRepoUrl, benchmarksBranch, _benchmarksDir);
+            var benchmarksDir = Git.Clone(path, _benchmarksRepoUrl);
```
Comment (halter73) [inline src/BenchmarksServer/Startup.cs]: Might it be simpler just to specify some random clone directory than read from stderr?

File: src/BenchmarksDriver/Program.cs [hunk 3]
```diff
@@ -74,12 +68,24 @@ public static int Main(string[] args)
...
-                        BenchmarksBranch = benchmarksBranchOption.Value(),
-                        BenchmarksRepo = benchmarksRepoOption.Value(),
-                        KestrelBranch = kestrelBranchOption.Value(),
-                        KestrelRepo = kestrelRepoOption.Value(),
                     };
 
+                    var sources = new List<Source>();
+                    foreach (var source in sourceOption.Values)
+                    {
+                        var split = source.LastIndexOf(':');
```
Comment (halter73) [inline src/BenchmarksDriver/Program.cs]: Use '@' as discussed.

File: src/BenchmarksServer/Startup.cs (hunk 1 above)
Comment (mikeharder) [inline src/BenchmarksServer/Startup.cs]: Will remove in next version of PR.

File: src/BenchmarksServer/Startup.cs (hunk 2 above)
Comment (mikeharder) [inline src/BenchmarksServer/Startup.cs]: It's nice to have meaningful directory names if you need to debug something, and the code to read from stderr is already written so it's easier to leave it at this point.

File: src/BenchmarksDriver/Program.cs [hunk 4]
```diff
@@ -46,16 +46,10 @@ public static int Main(string[] args)
...
-            var benchmarksRepoOption = app.Option("--benchmarksRepo",
-                "URL of Benchmarks repo.  Default is 'https://github.com/aspnet/benchmarks.git'.", CommandOptionType.SingleValue);
-
-            var kestrelBranchOption = app.Option("--kestrelBranch",
-                "Kestrel branch.  If specified, Benchmarks is configured to use Kestrel from sources rather than packages.",
-                CommandOptionType.SingleValue);
-            var kestrelRepoOption = app.Option("--kestrelRepo",
-                "URL of Kestrel repo.  Default is 'https://github.com/aspnet/KestrelHttpServer.git'.", CommandOptionType.SingleValue);
+            var sourceOption = app.Option("-o|--source",
+                "Source dependency. Format is 'repo:branch'. " +
```
Comment (halter73) [inline src/BenchmarksDriver/Program.cs]: Update help text to use '@'.

**Comment by halter73** [comment]
:shipit: after help text update.

**Pull request closed without merging**